storage-node-v2/src/services/sync/synchronizer.ts

```diff
@@ -0,0 +1,164 @@
+ * the source URL is resolved for each data object separately using the Query
+ * Node information about the storage providers.
+ */
+export async function performSync(
+  api: ApiPromise | undefined,
+  workerId: number,
+  asyncWorkersNumber: number,
+  queryNodeUrl: string,
+  uploadDirectory: string,
+  tempDirectory: string,
+  operatorUrl?: string
+): Promise<void> {
+  logger.info('Started syncing...')
+  const [model, files] = await Promise.all([
+    getStorageObligationsFromRuntime(queryNodeUrl, workerId),
+    getDataObjectIDs(),
+  ])
+
+  const requiredIds = model.dataObjects.map((obj) => obj.id)
+
+  const added = _.difference(requiredIds, files)
+  const deleted = _.difference(files, requiredIds)
+
+  logger.debug(`Sync - added objects: ${added.length}`)
+  logger.debug(`Sync - deleted objects: ${deleted.length}`)
+
+  const workingStack = new WorkingStack()
+  const deletedTasks = deleted.map((fileName) => new DeleteLocalFileTask(uploadDirectory, fileName))
+
+  let addedTasks: SyncTask[]
+  if (operatorUrl === undefined) {
+    addedTasks = await getPrepareDownloadTasks(
+      api,
+      model,
+      workerId,
+      added,
+      uploadDirectory,
+      tempDirectory,
+      workingStack
+    )
+  } else {
+    addedTasks = await getDownloadTasks(operatorUrl, added, uploadDirectory, tempDirectory)
+  }
+
+  logger.debug(`Sync - started processing...`)
+
+  const processSpawner = new TaskProcessorSpawner(workingStack, asyncWorkersNumber)
+
+  await workingStack.add(addedTasks)
+  await workingStack.add(deletedTasks)
+
+  await processSpawner.process()
+  logger.info('Sync ended.')
+}
+
+/**
+ * Creates the download preparation tasks.
+ *
+ * @param api - Runtime API promise
+ * @param dataObligations - defines the current data obligations for the node
+ * @param addedIds - data object IDs to download
+ * @param uploadDirectory - local directory for data uploading
+ * @param tempDirectory - local directory for temporary data uploading
+ * @param taskSink - a destination for the newly created tasks
+ */
+async function getPrepareDownloadTasks(
+  api: ApiPromise | undefined,
+  dataObligations: DataObligations,
+  currentWorkerId: number,
+  addedIds: string[],
+  uploadDirectory: string,
+  tempDirectory: string,
+  taskSink: TaskSink
+): Promise<PrepareDownloadFileTask[]> {
+  const bagIdByDataObjectId = new Map()
+  for (const entry of dataObligations.dataObjects) {
+    bagIdByDataObjectId.set(entry.id, entry.bagId)
+  }
+
+  const bucketOperatorUrlById = new Map()
+  for (const entry of dataObligations.storageBuckets) {
+    // Skip all buckets of the current WorkerId (this storage provider)
+    if (entry.workerId !== currentWorkerId) {
+      bucketOperatorUrlById.set(entry.id, entry.operatorUrl)
+    }
+  }
+
+  const bagOperatorsUrlsById = new Map()
+  for (const entry of dataObligations.bags) {
+    const operatorUrls = []
+
+    for (const bucket of entry.buckets) {
```

Comment:
I introduced filtering on the graphql queries level since we're moving towards safer data deletion. I think we shouldn't implement autocleaning for the metadata because this action won't be originated and reflected in the runtime.